### PR TITLE
wb_dcache: Forward "atomic transactions" to AXI

### PR DESCRIPTION
**Pull request by colluca**
### Description

This PR removes handling of atomic operations inside the WB cache. It forwards them instead on the AXI bypass interface, to be handled closer to memory. 

_Swap_, _fetch-and-op_ and _op_ type atomics are presented on the interface as AXI5 "Atomic Transactions".
_Load-reserved_ and _store-conditional_ are presented on the interface as AXI4 "Exclusive accesses".
This is aligned to how atomics are expected in the [_riscv_axi_atomics_](https://github.com/pulp-platform/axi_riscv_atomics) module.

### Status

- [x] **Forward _swap_, _fetch-and-op_ and _op_ type atomics to AXI interface**
Miss-hander FSM updated. AXI adapter extended to present AXI5 "Atomic transactions" requests and handle respective responses. 
- [x]  **Adapt testbench**
A _riscv_axi_atomics_ module is instantiated and connected between the DUT and the TB memory to handle atomics. The ensemble of the two modules should equal the previous DUT in terms of functionality, hence the rest of the TB can remain unaltered.

- [x] **Pass tests with _swap_, _fetch-and-op_ and _op_ type atomics**

- [x] **Forward _load-reserved_ and _store-conditional_ to AXI interface**
Extend AXI adapter to present AXI4 "Exclusive access" requests and handle respective responses. 

- [x] **Pass tests with _load-reserved_ and _store-conditional_ atomics**

### CODE REVIEW

**Comment by niwis** [comment]
Hi Luca, thanks a lot, looks good to me in general.

About the failing CI, it looks like the new `amo_i` port of `axi_adapter` still needs to be added in a few instantiations throughout the repo, e.g.
https://github.com/openhwgroup/cva6/blob/e748564dd880d68a9b0bde7e1a92c48c75823391/corev_apu/tb/ariane_testharness.sv#L283-L305
and
https://github.com/openhwgroup/cva6/blob/e748564dd880d68a9b0bde7e1a92c48c75823391/corev_apu/fpga/src/ariane_xilinx.sv#L552-L574

**Comment by colluca** [comment]
I made the corrections suggested, let's see if the CI passes now.

I had also added some TODOs in the code, which are some doubts I was left with. I think it might be good if you can have a look at them (or together in case) and then I can remove them before merging.

**Comment by colluca** [comment]
The previous commit corrects a bug when handling load atomics.

`valid_o` has two overlapping meanings. It signals an `rvalid` when handling read requests, and `bvalid` when handling write requests.
AXI5 "load" atomics, are those atomic operations which perform both a write and a read operation atomically. These employ the AW, W, B and R channels. In this case the `valid_o` output must be asserted with the latest between `bvalid` and `rvalid`.

For simplicity, we prioritize the B channel w.r.t to the R channel, i.e. we only accept (raise the corresponding ready signal) the R transaction after (or earliest together with) the B transaction.

**Comment by colluca** [comment]
The last commit reverts the AXI ID of the AMO port to 4'b1100. This value is the same as the AXI ID of the D$ port.
All the CI tests pass with this value.

With the value 4'b1011 which would be unique to the AMO port, the "ci / Atomic ISA Tests" fail.

File: core/axi_adapter.sv
Comment (niwis) [inline core/axi_adapter.sv:293]: I agree that we need to check the b response. But I don't think that it is guaranteed for the b response to arrive before the r response

**Review by niwis — CHANGES_REQUESTED** [review]
Hi Luca, I think we have to stick to `4'b10xx`. Otherwise, the responses will be routed to the dcache, which is surely wrong. I think we should not blindly trust the CI in this matter; better debug this in RTL simulation.
I also noticed a few potential issues (see review).

By the way, did you try the modular testbench for the latest changes?

Finally, please rebase the PR onto the latest master. This would help a lot with the Linux regression test :-)

File: core/axi_adapter.sv
Comment (colluca) [reply core/axi_adapter.sv:293]: Indeed not. But I don't acknowledge the r response until the b response has arrived. So we can assume it stays valid and pending until the b response is also received.

I believe this is the simplest implementation, otherwise we would have to at least register the r response.

File: core/cache_subsystem/miss_handler.sv
Comment (colluca) [reply core/cache_subsystem/miss_handler.sv:175]: I definitely agree. And I used `4'b1011` in Occamy and it doesn't pass the tests otherwise.

The question then is why the CI passes when the ID is `4'b1100`?

File: core/axi_adapter.sv
Comment (colluca) [reply core/axi_adapter.sv:419]: I double checked and you're right.

BTW, it's not an error I guess cause it gets truncated, but at line 211 in the `wt_axi_adapter` the concatenation operator is not really necessary. And it would actually be wrong to have the two MSBs set to `axi_pkg::ATOP_ATOMICLOAD`, so I guess that should be corrected.

File: core/cache_subsystem/miss_handler.sv
Comment (niwis) [reply core/cache_subsystem/miss_handler.sv:175]: Not sure, would need to run an RTL simulation to find out. But in general, false positives are possible.

File: core/axi_adapter.sv
Comment (niwis) [reply core/axi_adapter.sv]: I agree!

Comment (niwis) [reply core/axi_adapter.sv]: By the way: perhaps it would improve readability and maintainability if we would also use the ATOP params from `axi_pkg` instead of hardcoding the conversion. What do you think?

Comment (colluca) [reply core/axi_adapter.sv]: I can do this. Don't think it improves readability, but for sure it would have been less error-prone.

**Review by niwis — CHANGES_REQUESTED** [review]
Linux boots on fpga 🥳 The PR is good to go from my side after the two open nitpicks have been addressed (whitespace + ATOP encoding)

File: core/cache_subsystem/miss_handler.sv
Comment (colluca) [reply core/cache_subsystem/miss_handler.sv:62]: This seems correct to me with the latest commits

File: core/axi_adapter.sv
Comment (colluca) [reply core/axi_adapter.sv]: Corrected this :+1: So we can wait for the results of the CI and then promote it to proper PR

**Review by niwis — APPROVED** [review]

File: core/axi_adapter.sv
Comment (zarubaf) [inline core/axi_adapter.sv:157]: Does that synthesize?

File: core/cache_subsystem/miss_handler.sv
Comment (zarubaf) [inline core/cache_subsystem/miss_handler.sv:520]: Nit
```suggestion
        for (id = 0; id < NR_PORTS; id++) begin
```

File: core/cache_subsystem/miss_handler.sv
Comment (zarubaf) [inline core/cache_subsystem/miss_handler.sv:517]: Nit:
```suggestion
        logic [$clog2(NR_BYPASS_PORTS)-1:0] id;
```

**Review by zarubaf — APPROVED** [review]
LGTM, thanks! Two small nits.

File: core/axi_adapter.sv
Comment (colluca) [reply core/axi_adapter.sv:157]: Citing https://sutherland-hdl.com/papers/2015-SNUG-SV_SVA-for-RTL-Designers_paper.pdf section 2.1:

> Synthesis compilers ignore immediate assertions

Comment (zarubaf) [reply core/axi_adapter.sv:157]: Okay, let's see. Unfortunately, experience shows, that there can be a lot of written things and standards. At the end of the day, each tool has its own (partial) implementation of it.

**Pull request merged**
